File: src/cli/commands/diff.ts
import { resolve } from 'node:path';
import { existsSync } from 'node:fs';
import chalk from 'chalk';
import { GitBridge } from '../../git/bridge.js';
import type { DiffScope, FileChange } from '../../git/types.js';
import { ParserRegistry } from '../../parser/registry.js';
import { computeSemanticDiff } from '../../parser/differ.js';
import { SemDatabase } from '../../storage/database.js';
import { formatTerminal } from '../formatters/terminal.js';
import { formatJson } from '../formatters/json.js';
import { formatPlaintext } from '../formatters/plain.js';
import { formatMarkdown } from '../formatters/markdown.js';
import { createDefaultRegistry } from '../../parser/plugins/index.js';
import { loadConfig, validateChanges, formatValidationResults } from './validate.js';

interface ParsedArgs {
  scope?: DiffScope | { type: 'fileCompare'; a: string; b: string };
  pathspecs: string[];
}

function parseGitDiffArgs(args: string[]): ParsedArgs {
  // Split on "--" separator
  const sepIdx = args.indexOf('--');
  let refs: string[];
  let pathspecs: string[];
  if (sepIdx >= 0) {
    refs = args.slice(0, sepIdx);
    pathspecs = args.slice(sepIdx + 1);
  } else {
    refs = args;
    pathspecs = [];
  }

  if (refs.length === 0) {
    return { pathspecs };
  }

  if (refs.length === 1) {
    const arg = refs[0];
    // Check for ... (merge-base) first
    const tripleIdx = arg.indexOf('...');
    if (tripleIdx > 0 && tripleIdx < arg.length - 3) {
      return {
        scope: { type: 'range', from: `${arg.slice(0, tripleIdx)}...mergebase`, to: arg.slice(tripleIdx + 3) },
        pathspecs,
      };
    }
    // Check for .. (range)
    const doubleIdx = arg.indexOf('..');
    if (doubleIdx > 0 && doubleIdx < arg.length - 2) {
      return {
        scope: { type: 'range', from: arg.slice(0, doubleIdx), to: arg.slice(doubleIdx + 2) },
        pathspecs,
      };
    }
    // Single ref
    return { scope: { type: 'refToWorking', refspec: arg }, pathspecs };
  }

  if (refs.length === 2) {
    // Check if both exist as files
    if (pathspecs.length === 0 && existsSync(refs[0]) && existsSync(refs[1])) {
      return { scope: { type: 'fileCompare', a: refs[0], b: refs[1] }, pathspecs };
    }
    return { scope: { type: 'range', from: refs[0], to: refs[1] }, pathspecs };
  }

  console.error(chalk.red('Error: too many positional arguments. Use -- to separate pathspecs.'));
  process.exit(1);
}

export interface DiffOptions {
  cwd?: string;
  format?: 'terminal' | 'json' | 'plain' | 'markdown';
  staged?: boolean;
  commit?: string;
  from?: string;
  to?: string;
  store?: boolean;
  args?: string[];
}

// Singleton registry — no need to recreate on every call
let _registry: ParserRegistry | undefined;
function getRegistry(): ParserRegistry {
  return (_registry ??= createDefaultRegistry());
}

export async function diffCommand(opts: DiffOptions = {}): Promise<void> {
  const cwd = opts.cwd ?? process.cwd();
  const git = new GitBridge(cwd);

  const parsed = parseGitDiffArgs(opts.args ?? []);

  let scope: DiffScope;
  let fileChanges: FileChange[];

  // Handle file comparison mode (two file paths)
  if (parsed.scope?.type === 'fileCompare') {
    const { a, b } = parsed.scope;
    // If in a git repo and both resolve as refs, prefer ref comparison
    if (await git.isRepo()) {
      const [aIsRef, bIsRef] = await Promise.all([git.isValidRev(a), git.isValidRev(b)]);
      if (aIsRef && bIsRef) {
        scope = { type: 'range', from: a, to: b };
        fileChanges = await git.getChangedFiles(scope);
        return runDiffPipeline(fileChanges, scope, git, opts);
      }
    }
    // Fall back to direct file comparison
    const { readFile } = await import('node:fs/promises');
    const beforeContent = await readFile(resolve(cwd, a), 'utf-8');
    const afterContent = await readFile(resolve(cwd, b), 'utf-8');
    fileChanges = [{
      filePath: b,
      status: 'modified',
      beforeContent,
      afterContent,
    }];
    scope = { type: 'working' };
    return runDiffPipeline(fileChanges, scope, git, opts);
  }

  // Determine scope from explicit flags, parsed args, or auto-detect
  if (opts.commit) {
    if (!(await git.isRepo())) { console.error(chalk.red('Error: Not inside a Git repository.')); process.exit(1); }
    scope = { type: 'commit', sha: opts.commit };
    fileChanges = await git.getChangedFiles(scope);
  } else if (opts.from && opts.to) {
    if (!(await git.isRepo())) { console.error(chalk.red('Error: Not inside a Git repository.')); process.exit(1); }
    scope = { type: 'range', from: opts.from, to: opts.to };
    fileChanges = await git.getChangedFiles(scope);
  } else if (parsed.scope) {
    if (!(await git.isRepo())) { console.error(chalk.red('Error: Not inside a Git repository.')); process.exit(1); }
    // Handle merge-base range (from contains "...mergebase" marker)
    if (parsed.scope.type === 'range' && parsed.scope.from.endsWith('...mergebase')) {
      const ref1 = parsed.scope.from.slice(0, -'...mergebase'.length);
      const ref2 = parsed.scope.to;
      try {
        const base = await git.resolveMergeBase(ref1, ref2);
        scope = { type: 'range', from: base, to: ref2 };
      } catch (e) {
        console.error(chalk.red(`Error resolving merge base: ${e}`));
        process.exit(1);
      }
    } else {
      scope = parsed.scope as DiffScope;
    }
    fileChanges = await git.getChangedFiles(scope);
  } else if (opts.staged) {
    if (!(await git.isRepo())) { console.error(chalk.red('Error: Not inside a Git repository.')); process.exit(1); }
    scope = { type: 'staged' };
    fileChanges = await git.getChangedFiles(scope);
  } else {
    // Combined: isRepo + detectScope + getChangedFiles in one batch
    try {
      const result = await git.detectAndGetFiles();
      scope = result.scope;
      fileChanges = result.files;
    } catch {
      console.error(chalk.red('Error: Not inside a Git repository.'));
      process.exit(1);
    }
  }

  // Filter by pathspecs if provided
  if (parsed.pathspecs.length > 0) {
    fileChanges = fileChanges.filter(f =>
      parsed.pathspecs.some(spec =>
        f.filePath === spec ||
        f.filePath.startsWith(spec.endsWith('/') ? spec : `${spec}/`)
      )
    );
  }

  return runDiffPipeline(fileChanges, scope!, git, opts);
}

async function runDiffPipeline(
  fileChanges: FileChange[],
  scope: DiffScope,
  git: GitBridge,
  opts: DiffOptions,
): Promise<void> {
  if (fileChanges.length === 0) {
    console.log(chalk.dim('No changes detected.'));
    return;
  }

  // Compute semantic diff
  const registry = getRegistry();
  const commitSha = scope.type === 'commit' ? scope.sha : undefined;
  const result = computeSemanticDiff(fileChanges, registry, commitSha);

  // Get repoRoot once for both store + validation
  let repoRoot: string;
  try {
    repoRoot = await git.getRepoRoot();
  } catch {
    repoRoot = opts.cwd ?? process.cwd();
  }

  // Optionally store changes
  if (opts.store) {
    const dbPath = resolve(repoRoot, '.sem', 'sem.db');
    if (existsSync(dbPath)) {
      const db = new SemDatabase(dbPath);
      db.insertChanges(result.changes);
      db.close();
    }
  }

  // Output
  const format = opts.format ?? 'terminal';
  if (format === 'json') {
    console.log(formatJson(result));
  } else if (format === 'plain') {
    console.log(formatPlaintext(result));
  } else if (format === 'markdown') {
    console.log(formatMarkdown(result));
  } else {
    console.log(formatTerminal(result));
  }

  // Run validation rules if .semrc exists
  try {
    const config = await loadConfig(repoRoot);
    if (config.rules && config.rules.length > 0) {
      const violations = validateChanges(result, config);
      if (violations.length > 0) {
        console.log('');
        console.log(formatValidationResults(violations));
      }
    }
  } catch {
    // No config or invalid config — skip validation
  }
}

File: crates/sem-cli/src/commands/diff.rs
use std::io::Read;
use std::path::Path;
use std::process;
use std::time::Instant;

use sem_core::git::bridge::GitBridge;
use sem_core::git::types::{DiffScope, FileChange};
use sem_core::parser::differ::compute_semantic_diff;
use sem_core::parser::plugins::create_default_registry;

use crate::formatters::{json::format_json, markdown::format_markdown, plain::format_plain, terminal::format_terminal};

pub struct DiffOptions {
    pub cwd: String,
    pub format: OutputFormat,
    pub staged: bool,
    pub commit: Option<String>,
    pub from: Option<String>,
    pub to: Option<String>,
    pub stdin: bool,
    pub verbose: bool,
    pub profile: bool,
    pub file_exts: Vec<String>,
    pub args: Vec<String>,
}

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum OutputFormat {
    Terminal,
    Plain,
    Json,
    Markdown,
}

/// Parsed result of git-diff-style positional arguments
struct ParsedArgs {
    /// The resolved diff scope (None = auto-detect)
    scope: Option<ParsedScope>,
    /// Pathspecs for filtering (after --)
    pathspecs: Vec<String>,
}

enum ParsedScope {
    /// Two files to compare directly
    FileCompare(String, String),
    /// A single ref compared to working tree
    RefToWorking(String),
    /// A range between two refs
    Range(String, String),
    /// A merge-base range (ref1...ref2)
    MergeBaseRange(String, String),
}

/// Split args on "--" separator into (refs_or_files, pathspecs)
fn split_on_separator(args: Vec<String>) -> (Vec<String>, Vec<String>) {
    if let Some(pos) = args.iter().position(|a| a == "--") {
        let mut args = args;
        let pathspecs = args.split_off(pos + 1);
        args.pop(); // remove the "--"
        (args, pathspecs)
    } else {
        (args, vec![])
    }
}

fn parse_args(args: Vec<String>) -> ParsedArgs {
    let (refs, pathspecs) = split_on_separator(args);

    if refs.is_empty() {
        return ParsedArgs { scope: None, pathspecs };
    }

    if refs.len() == 1 {
        let arg = &refs[0];

        // Check for ... (merge-base) syntax first (before ..)
        if let Some((from, to)) = arg.split_once("...") {
            if !from.is_empty() && !to.is_empty() {
                return ParsedArgs {
                    scope: Some(ParsedScope::MergeBaseRange(from.to_string(), to.to_string())),
                    pathspecs,
                };
            }
        }

        // Check for .. (range) syntax
        if let Some((from, to)) = arg.split_once("..") {
            if !from.is_empty() && !to.is_empty() {
                return ParsedArgs {
                    scope: Some(ParsedScope::Range(from.to_string(), to.to_string())),
                    pathspecs,
                };
            }
        }

        // Single ref → compare to working tree
        return ParsedArgs {
            scope: Some(ParsedScope::RefToWorking(arg.clone())),
            pathspecs,
        };
    }

    if refs.len() == 2 {
        let a = &refs[0];
        let b = &refs[1];

        // If both exist as files on disk and no pathspecs, treat as file comparison
        if pathspecs.is_empty() && Path::new(a).exists() && Path::new(b).exists() {
            // But check if they're also valid git refs — prefer ref interpretation
            // Only fall back to file comparison if neither resolves as a ref
            return ParsedArgs {
                scope: Some(ParsedScope::FileCompare(a.clone(), b.clone())),
                pathspecs,
            };
        }

        // Two refs → range
        return ParsedArgs {
            scope: Some(ParsedScope::Range(a.clone(), b.clone())),
            pathspecs,
        };
    }

    eprintln!("\x1b[31mError: too many positional arguments. Use -- to separate pathspecs.\x1b[0m");
    process::exit(1);
}

pub fn diff_command(mut opts: DiffOptions) {
    let total_start = Instant::now();

    let t0 = Instant::now();
    let parsed = parse_args(std::mem::take(&mut opts.args));

    let (file_changes, from_stdin) = if opts.stdin {
        // Read FileChange[] from stdin — no git repo needed
        let mut input = String::new();
        std::io::stdin().read_to_string(&mut input).unwrap_or_else(|e| {
            eprintln!("\x1b[31mError reading stdin: {e}\x1b[0m");
            process::exit(1);
        });
        let changes: Vec<FileChange> = serde_json::from_str(&input).unwrap_or_else(|e| {
            eprintln!("\x1b[31mError parsing stdin JSON: {e}\x1b[0m");
            process::exit(1);
        });
        (changes, true)
    } else if let Some(ParsedScope::FileCompare(ref a, ref b)) = parsed.scope {
        // Compare two arbitrary files: sem diff file1.ts file2.ts
        let path_a = Path::new(a);
        let path_b = Path::new(b);

        // If we're in a git repo and both resolve as refs, prefer ref comparison
        if let Ok(git) = GitBridge::open(Path::new(&opts.cwd)) {
            if git.is_valid_rev(a) && git.is_valid_rev(b) {
                let scope = DiffScope::Range { from: a.clone(), to: b.clone() };
                match git.get_changed_files(&scope, &parsed.pathspecs) {
                    Ok(files) => return run_diff_pipeline(files, false, &opts, &parsed, total_start, t0),
                    Err(e) => {
                        eprintln!("\x1b[31mError: {e}\x1b[0m");
                        process::exit(1);
                    }
                }
            }
        }

        let content_a = std::fs::read_to_string(path_a).unwrap_or_else(|e| {
            eprintln!("\x1b[31mError reading {}: {e}\x1b[0m", path_a.display());
            process::exit(1);
        });
        let content_b = std::fs::read_to_string(path_b).unwrap_or_else(|e| {
            eprintln!("\x1b[31mError reading {}: {e}\x1b[0m", path_b.display());
            process::exit(1);
        });

        let change = FileChange {
            file_path: b.clone(),
            old_file_path: None,
            status: sem_core::git::types::FileStatus::Modified,
            before_content: Some(content_a),
            after_content: Some(content_b),
        };
        (vec![change], false)
    } else {
        let git = match GitBridge::open(Path::new(&opts.cwd)) {
            Ok(g) => g,
            Err(_) => {
                eprintln!("\x1b[31mError: Not inside a Git repository.\x1b[0m");
                process::exit(1);
            }
        };

        // Determine scope from explicit flags, parsed args, or auto-detect
        let (_scope, file_changes) = if let Some(ref sha) = opts.commit {
            let scope = DiffScope::Commit { sha: sha.clone() };
            match git.get_changed_files(&scope, &parsed.pathspecs) {
                Ok(files) => (scope, files),
                Err(e) => {
                    eprintln!("\x1b[31mError: {e}\x1b[0m");
                    process::exit(1);
                }
            }
        } else if let (Some(ref from), Some(ref to)) = (&opts.from, &opts.to) {
            let scope = DiffScope::Range {
                from: from.clone(),
                to: to.clone(),
            };
            match git.get_changed_files(&scope, &parsed.pathspecs) {
                Ok(files) => (scope, files),
                Err(e) => {
                    eprintln!("\x1b[31mError: {e}\x1b[0m");
                    process::exit(1);
                }
            }
        } else if let Some(ref parsed_scope) = parsed.scope {
            // Use scope from positional args
            let scope = match parsed_scope {
                ParsedScope::RefToWorking(refspec) => {
                    if opts.staged {
                        // git diff --cached <ref> = compare ref to index
                        // We approximate this as Range from ref to HEAD (staged view)
                        // For now, just use the ref as a range base
                        DiffScope::Range {
                            from: refspec.clone(),
                            to: "HEAD".to_string(),
                        }
                    } else {
                        DiffScope::RefToWorking { refspec: refspec.clone() }
                    }
                }
                ParsedScope::Range(from, to) => DiffScope::Range {
                    from: from.clone(),
                    to: to.clone(),
                },
                ParsedScope::MergeBaseRange(ref1, ref2) => {
                    match git.resolve_merge_base(ref1, ref2) {
                        Ok(base) => DiffScope::Range {
                            from: base,
                            to: ref2.clone(),
                        },
                        Err(e) => {
                            eprintln!("\x1b[31mError resolving merge base: {e}\x1b[0m");
                            process::exit(1);
                        }
                    }
                }
                ParsedScope::FileCompare(_, _) => unreachable!(),
            };
            match git.get_changed_files(&scope, &parsed.pathspecs) {
                Ok(files) => (scope, files),
                Err(e) => {
                    eprintln!("\x1b[31mError: {e}\x1b[0m");
                    process::exit(1);
                }
            }
        } else if opts.staged {
            let scope = DiffScope::Staged;
            match git.get_changed_files(&scope, &parsed.pathspecs) {
                Ok(files) => (scope, files),
                Err(e) => {
                    eprintln!("\x1b[31mError: {e}\x1b[0m");
                    process::exit(1);
                }
            }
        } else {
            match git.detect_and_get_files(&parsed.pathspecs) {
                Ok((scope, files)) => (scope, files),
                Err(_) => {
                    eprintln!("\x1b[31mError: Not inside a Git repository.\x1b[0m");
                    process::exit(1);
                }
            }
        };
        (file_changes, false)
    };

    run_diff_pipeline(file_changes, from_stdin, &opts, &parsed, total_start, t0);
}

fn run_diff_pipeline(
    file_changes: Vec<FileChange>,
    from_stdin: bool,
    opts: &DiffOptions,
    _parsed: &ParsedArgs,
    total_start: Instant,
    t0: Instant,
) {
    let git_diff_ms = t0.elapsed().as_secs_f64() * 1000.0;

    // Filter by file extensions if specified
    let file_changes = if opts.file_exts.is_empty() {
        file_changes
    } else {
        let exts: Vec<String> = opts.file_exts.iter().map(|e| {
            if e.starts_with('.') { e.clone() } else { format!(".{}", e) }
        }).collect();
        file_changes.into_iter().filter(|fc| {
            exts.iter().any(|ext| fc.file_path.ends_with(ext.as_str()))
        }).collect()
    };

    if file_changes.is_empty() {
        println!("\x1b[2mNo changes detected.\x1b[0m");
        return;
    }

    let t2 = Instant::now();
    let registry = create_default_registry();
    let registry_ms = t2.elapsed().as_secs_f64() * 1000.0;

    let t3 = Instant::now();
    let result = compute_semantic_diff(&file_changes, &registry, None, None);
    let parse_diff_ms = t3.elapsed().as_secs_f64() * 1000.0;

    let t4 = Instant::now();
    let output = match opts.format {
        OutputFormat::Json => format_json(&result),
        OutputFormat::Markdown => format_markdown(&result),
        OutputFormat::Plain => format_plain(&result),
        OutputFormat::Terminal => format_terminal(&result, opts.verbose),
    };
    let format_ms = t4.elapsed().as_secs_f64() * 1000.0;

    println!("{output}");

    if opts.profile {
        let total_ms = total_start.elapsed().as_secs_f64() * 1000.0;
        eprintln!();
        eprintln!("\x1b[2m── Profile ──────────────────────────────────\x1b[0m");
        eprintln!("\x1b[2m  input ({})  {git_diff_ms:>8.2}ms\x1b[0m",
            if from_stdin { "stdin" } else { "git" });
        eprintln!("\x1b[2m  registry init        {registry_ms:>8.2}ms\x1b[0m");
        eprintln!("\x1b[2m  parse + match        {parse_diff_ms:>8.2}ms\x1b[0m");
        eprintln!("\x1b[2m  format output        {format_ms:>8.2}ms\x1b[0m");
        eprintln!("\x1b[2m  ─────────────────────────────────────────────\x1b[0m");
        eprintln!("\x1b[2m  total                {total_ms:>8.2}ms\x1b[0m");
        eprintln!("\x1b[2m  files: {}  entities: {}  changes: {}\x1b[0m",
            file_changes.len(), result.changes.len(),
            result.added_count + result.modified_count + result.deleted_count + result.moved_count + result.renamed_count);
        eprintln!("\x1b[2m─────────────────────────────────────────────\x1b[0m");
    }
}
